src/Microsoft.OData.Core/JsonLight/ODataJsonLightPropertySerializer.cs

```diff
@@ -227,9 +246,9 @@ private bool IsOpenProperty(ODataProperty property)
                 return;
             }
 
-            if (value is ODataNullValue || value == null)
+            if (isNullValue)
             {
-                this.WriteNullProperty(property);
+                this.WriteNullProperty(property, omitNullValues);

```

Comment:
do you need the new omitNullValues parameter in WriteNullProperty? Why would you get this far if the property was a null value and omitNullValues was true?  Is this only for validating the metadata?  Can you handle that when you check for isNullValue, and get rid of the new omitNullValues flag to WriteNullProperty?

Reply:
The omitNullValues is needed in WriteNullProperty to differentiate delta and others; When we get this far, we knows property is a null but omitNullValues can still be true or false; I think validating metadata is done before this (L215,L217); I believe WriteNullProperty properly encapsulates the logic to write null property, without the omitNullValues flag it wouldn't be able to tell whether null values needs to be serialized (e.g. delata or non-delta).
I have added additional information about the omitNullValues param in pending changes.

---
In reply to: [186561856](https://github.com/OData/odata.net/pull/1106#discussion_r186561856) [](ancestors = 186561856)

Reply:
On L219, if the value is the null value and omitNullValues is true, you would have returned, so you would never reach this line for a null value if omitNullValues is true.  Right?

---
In reply to: [187209557](https://github.com/OData/odata.net/pull/1106#discussion_r187209557) [](ancestors = 187209557,186561856)

Reply:
There is a nested if-clause inside L219, code won't return when the if-clause @L221 is not satisfied.

---
In reply to: [198078564](https://github.com/OData/odata.net/pull/1106#discussion_r198078564) [](ancestors = 198078564,187209557,186561856)

Reply:
The nested if clause checks two things:
1) If this is top level, in which case omitNullValues doesn't apply (it's already ignored in WriteNullProperty)
2) whether we should check metadata

For #2, I'm not convinced we should skip checking metadata for null values as this would introduce a change in what the writer was able to write whether omit values was true or false:
a) this would allow the writer to write a non-null value and one or more null values for the same property when omitNullValues was true, which should really not be valid.
b) this would allow the writer to write a null value for an undeclared property, even though throwOnUndeclaredProperty was true, when omitNullValues is true

So, we should simplify move the validations in L227 and L229 before the check for null values in L219, and in L219 return for null non-top level elements if omitNullValues is true.  Then we can remove the omitNullValues argument from WriteNullProperty, since it will only be called when we really want to write the null value.

---
In reply to: [209787942](https://github.com/OData/odata.net/pull/1106#discussion_r209787942) [](ancestors = 209787942,198078564,187209557,186561856)

Reply:
For performance reason, kosinsky has requested (see his comment at L192) that skipping the validations at L227-229, using check for null values in L219 *before hand*. Switching the order as you suggest above would contradict that, right?

---
In reply to: [213515156](https://github.com/OData/odata.net/pull/1106#discussion_r213515156) [](ancestors = 213515156,209787942,198078564,187209557,186561856)

Reply:
Quite honestly, throwOnUndeclaredProperty is a carry-over from OData V3.  In OData V4 it should always be false which means this check should never have to occur.  Unfortunately, the default value is currently false, but we should consider changing it.

I still dislike having the behavior change based on the omit-nulls flag; having the same code throw an exception if omit-nulls is false and succeed if omit-nulls is true is unintuitive; for example, avoiding the duplicateNamesChecker means that the app could succeed in writing the same property twice, once as null and once not as null, with omit-nulls equal to true, and fail with omit-nulls equal to false. Omit-nulls should we a wire optimization, not a change in behavior.

All that said, it's a nice optimization, even if throwOnUndeclaredProperty is false, to avoid the duplicateNameChecker....

---
In reply to: [213868365](https://github.com/OData/odata.net/pull/1106#discussion_r213868365) [](ancestors = 213868365,213515156,209787942,198078564,187209557,186561856)

Reply:
As we discussed with kosinsky, we can ignore the optimization for validation. I have updated the code accordingly including removal of the omitNullValues argument from WriteNullProperty. Thanks.

---
In reply to: [229533703](https://github.com/OData/odata.net/pull/1106#discussion_r229533703) [](ancestors = 229533703,213868365,213515156,209787942,198078564,187209557,186561856)